### ISSUE TITLE
Issue with use of null char (\x00) in filter string (ettercap 0.8.3.1)?

### ISSUE THREAD

**Issue by graythang** [Ettercap/ettercap#1087]
Is there a known issue with use of the null hex code in filters?  It appears to be truncating the string whenever a \x00 is encountered.  Here is what Im seeing:

Test case using \x00 in string (second test case below is w/o the null in the string)

************ test case 1 using \x00*************
 >etterfilter myfilter 
     where myfilter contains:

if (ip.proto == TCP && tcp.dst == 1433) {
        msg("Found Our target Traffic");
        if (search(DATA.data,"\x61\x00\x61\x00\x61\x00\x61\x00\x74\x00\x65\x00\x73\x00\x74\x00\x31\x00\x31")) {
        msg("Found Our target String");
        replace("\x61\x00\x61\x00\x61\x00\x61\x00\x74\x00\x65\x00\x73\x00\x74\x00\x31\x00\x31\x00","\x62\x00\x62\x00\x62\x00\x62\x00\x74\x00\x65\x00\x73\x00\x74\x00\x31\x00\x31\x00");
        msg("...and replaced it:)");
        }
}

************
etterfilter -t filter.ef   results in: 
etterfilter 0.8.3.1 copyright 2001-2020 Ettercap Development Team

Content filters loaded from filter.ef...
Disassebling "filter.ef" content...

0000: TEST level 3, offset 9, size 1, == 6 [0x6]
0001: JUMP IF FALSE to 0010
0002: TEST level 4, offset 2, size 2, == 1433 [0x599]
0003: JUMP IF FALSE to 0010
0004: MSG "Found Our target Traffic"
0005: SEARCH level 5, string "a"
0006: JUMP IF FALSE to 0010
0007: MSG "Found Our target String"
0008: REPLACE "a" --> "b"
0009: MSG "...and replaced it:)"

 10 instructions decoded.

***************** test case 2 no \x00 string *********************
Second test case omitting "\x00" from filter string
 >etterfilter myfilter2 
     where myfilter2 contains:
if (ip.proto == TCP && tcp.dst == 1433) {
        msg("Found Our target Traffic");
        if (search(DATA.data,"\x61\x61\x61\x61\x74\x65\x73\x74\x31\x31")) {
        msg("Found Our target String");
        replace("\x61\x61\x61\x61\x74\x65\x73\x74\x31\x31","\x62\x62\x62\x62\x74\x65\x73\x74\x31\x31");
        msg("...and replaced it:)");
        }
}

etterfilter -t on filter.ef results in
etterfilter 0.8.3.1 copyright 2001-2020 Ettercap Development Team

Content filters loaded from filter.ef...
Disassebling "filter.ef" content...

0000: TEST level 3, offset 9, size 1, == 6 [0x6]
0001: JUMP IF FALSE to 0010
0002: TEST level 4, offset 2, size 2, == 1433 [0x599]
0003: JUMP IF FALSE to 0010
0004: MSG "Found Our target Traffic"
0005: SEARCH level 5, string "aaaatest11"
0006: JUMP IF FALSE to 0010
0007: MSG "Found Our target String"
0008: REPLACE "aaaatest11" --> "bbbbtest11"
0009: MSG "...and replaced it:)"

 10 instructions decoded.

**Comment by koeppea** [comment]
There have been a zero-byte string termination issue but as far as I remember, I've fixed this, and as far as I remember it should also be fixed in version 0.8.3.1.
Could you please provide a packet capture file which I can use for testing your filter?

**Comment by graythang** [comment]
Unfortunately the traffic is from a penetration test and can't be shared due to contractual obligations.  But given the etterfilter -t shows the string was truncated wouldn't the first step be to move past that before testing with a pcap could even begin?  In the example I provided the "aaaatest11" and "bbbbtest11" were used with one run including the null characters between ascii characters (MSSQL uses this format in the TDS packets), while the other didn't.  the etterfilter -t dump shows the one using Null's was truncated.

**Comment by LocutusOfBorg** [comment]
```
      USER_MSG("%04lu: ASSIGNMENT level %d, offset %d, string \"%s\"\n", (unsigned long)eip,
            fop->op.assign.level, fop->op.assign.offset, fop->op.assign.string);
```
printing 00 with %s leads to this problem...
```
         u_int8   *string;
         size_t   slen;
```
We should print with slen but I'm not sure slen is correctly feeded

**Comment by LocutusOfBorg** [comment]
https://stackoverflow.com/questions/6093660/printf-with-s-to-include-null-characters

**Comment by koeppea** [comment]
Well I don't think that it's related to the `printf()` issue, even though it's true.
I assume it's related due to the use of str* functions which also early stop at the first null-byte.
The commit 4eacdca1b8b3ea45107e1ca1966a282a6cf82058 I was referring to, fixed it only for the assignment and comparison operators.

However, I'm a bit cautious making replace capable replacing binary strings, since I'm afraid that might break existing filters.
Maybe a new variant named differently. Maybe `raw_replace()`.
@LocutusOfBorg what is your opinion?

**Comment by LocutusOfBorg** [comment]
I agree on this, but please if we calloc, we should also SAFE_FREE :) #942

**Comment by koeppea** [comment]
Well of course. 

But when I replace a strdup with calloc/memcpy, I normally doesn’t specifically check if the prior strdup allocation is freed somewhere. 

But you’re absolutely right. What is allocated must be freed.

**Comment by koeppea** [comment]
I've now managed to reproduce the scenario.
Took over filter source 1:1
```
$ cat mssql-replace.ecf 
if (ip.proto == TCP && tcp.dst == 1433) {
   msg("Found Our target Traffic");
      if (search(DATA.data,"\x61\x00\x61\x00\x61\x00\x61\x00\x74\x00\x65\x00\x73\x00\x74\x00\x31\x00\x31")) {
         msg("Found Our target String");
         replace("\x61\x00\x61\x00\x61\x00\x61\x00\x74\x00\x65\x00\x73\x00\x74\x00\x31\x00\x31\x00","\x62\x00\x62\x00\x62\x00\x62\x00\x74\x00\x65\x00\x73\x00\x74\x00\x31\x00\x31\x00");
         msg("...and replaced it:)");
   }
}
$
```
Generated binary payload file from the filter byte pattern:
```
$ hd mssql.payload 
00000000  61 00 61 00 61 00 61 00  74 00 65 00 73 00 74 00  |a.a.a.a.t.e.s.t.|
00000010  31 00 31 00                                       |1.1.|
00000014
$
```
Ettercap session with the compiled filter loaded.
```
Host 172.21.21.1 added to TARGET1
Host 172.21.21.34 added to TARGET2

ARP poisoning victims:

 GROUP 1 : 172.21.21.1 E0:28:6D:47:C2:19

 GROUP 2 : 172.21.21.34 08:00:27:7E:2C:B7
Content filters loaded from /tmp/mssql-replace.ef...
Found Our target Traffic
Found Our target String
...and replaced it:)
```
Here is the Wireshark packet details of the received and modified forwarded packet:

![grafik](https://user-images.githubusercontent.com/4201853/100485018-d9de8f00-30fe-11eb-8b46-09e821710cc6.png)


Conclusion: I wasn't able to reproduce. In my case, the string has been replaced as expected.
